### DISCUSSION TITLE
chatGPT knows py5! (although not so good)

### DISCUSSION THREAD

**Discussion by Maltiec** [py5coding/py5generator#291] [General]
Some days ago my student showed me sketch in py5 made with a help of chatGPT, I've decided to make some experiments too. This one looks nice:
```python
import py5

def setup():
    py5.size(400, 400)

def draw():
    py5.background(255)
    x = py5.mouse_x
    y = py5.mouse_y
    d = py5.dist(py5.width / 2, py5.height / 2, x, y)
    # Calculating circle color
    r = py5.remap(x, 0, py5.width, 0, 255)
    g = py5.remap(y, 0, py5.height, 0, 255)
    b = py5.remap(d, 0, py5.width / 2, 255, 0)
    py5.fill(r, g, b)
    py5.stroke(0)
    py5.ellipse(x, y, 50, 50)

py5.run_sketch()

``` 

Results:
1. chatGPT knows py5 (though in above example it made a mistake and used py5.map instead of py5.remap, I've made correction myself)
2. For now if I try to tell it to use objects and py5, it tells that doesn't know how to code with py5 and shows the alternative with pygame.
I'll continue my experiments. )

**Comment by Maltiec** [comment]
update - it can use objects if I insist on using class and give a name for it:
```python
import py5

class myCircle:
    def __init__(self, x, y, size=50):
        self.x = x
        self.y = y
        self.size = size
        self.color = (0, 0, 0)
    
    def display(self):
        py5.fill(*self.color)
        py5.circle(self.x, self.y, self.size)
    
    def move(self, dx, dy):
        self.x += dx
        self.y += dy
        self.color = (py5.map(self.x, 0, py5.width, 0, 255), 
                      py5.map(self.y, 0, py5.height, 0, 255), 
                      0)

circle = myCircle(py5.width // 2, py5.height // 2)

def setup():
    py5.size(400, 400)

def draw():
    py5.background(255)
    circle.display()
    circle.move(py5.random_uniform(-5, 5), py5.random_uniform(-5, 5))

py5.run_sketch()
```

**Reply by hx2A** [reply]
Wow, I'm surprised ChatGPT does this well with py5 coding! My understanding is that ChatGPT's dataset is up until sometime in 2021, which is the same year py5 became publicly available. I don't remember exactly when that happened, I think it was around March or April of 2021. In any case, there must be few if any examples of actual py5 code in ChatGPT's training dataset.

It's interesting to compare/contrast how ChatGPT does for the JavaScript library p5 and the Python library py5. Conceptually they are very similar but ChatGPT must have seen a lot of p5 examples in its training set. If ChatGPT can code both equally well, it suggests it is able to extend its understanding of p5 to py5. It isn't just regurgitating stuff from its training set.

**Reply by Maltiec** [reply]
Py5Vector class was in the very first release of py5? I am trying to insist on using vectors to move objects, but chat GPT says it can't code in Python and gives just the basic concept of such application. ) Btw, it generated code for the same prompt when I changed python to js and py5 to p5

**Reply by hx2A** [reply]
No, Py5Vector didn't appear until 0.7.0.

https://github.com/py5coding/py5generator/releases/tag/py5-1277-0012-0.7.0a0

> Btw, it generated code for the same prompt when I changed python to js and py5 to p5

Hmmm, so perhaps ChatGPT has some ability to extend its understanding of p5 to py5, but it is not complete.

I wonder how ChatGPT will impact the libraries and code people use? I imagine that libraries that ChatGPT can't write good code for will be used less. Will this start a cycle, where they continue to get used less often, and others that it does know get used more? I have concerns this will have a negative impact on the open source world's ability to innovate.

**Reply by hx2A** [reply]
BTW, there is a py5 release today!

https://github.com/py5coding/py5generator/releases/tag/py5-1292-0019-0.9.0a0

**Reply by Maltiec** [reply]
> No, Py5Vector didn't appear until 0.7.0.
> 
> https://github.com/py5coding/py5generator/releases/tag/py5-1277-0012-0.7.0a0
> 
> > Btw, it generated code for the same prompt when I changed python to js and py5 to p5
> 
> Hmmm, so perhaps ChatGPT has some ability to extend its understanding of p5 to py5, but it is not complete.
> 
> I wonder how ChatGPT will impact the libraries and code people use? I imagine that libraries that ChatGPT can't write good code for will be used less. Will this start a cycle, where they continue to get used less often, and others that it does know get used more? I have concerns this will have a negative impact on the open source world's ability to innovate.

Well, further experiments show that it is a problem of understanding by network, so after I told her to evaluate her code by using Py5Vector class she (ehh, I am a bit of AI optimist so sometimes I see a personality inside ) ) made refactoring, and used vector class from Processing Python mode. This is chatGPT3.5, I'm going to try v.4 - may be it'll work better. And also there is a fine tuning available - may be some new training data could be injected into network.

**Reply by Maltiec** [reply]
update: chatGPT4 is much better at py5 programming. Either dataset was updated, or she is really can "think ahead" when lacks examples.

**Reply by hx2A** [reply]
Excellent. I'm glad ChatGPT4 can write solid py5 code. That can only help encourage new users.
